In [1]:
import pandas as pd
import numpy as np
import os

In [2]:
# storing the current working directory path
cwd = os.getcwd()

In [3]:
# reading the companies text file
companies = pd.read_csv('companies.txt',sep="\t",encoding='unicode_escape')

In [4]:
companies.head()

permalink                    name  \
0                  /Organization/-Fame                   #fame   
1               /Organization/-Qounter                :Qounter   
2  /Organization/-The-One-Of-Them-Inc-  (THE) ONE of THEM,Inc.   
3                /Organization/0-6-Com                 0-6.com   
4       /Organization/004-Technologies        004 Technologies   

                        homepage_url  \
0                 http://livfame.com   
1             http://www.qounter.com   
2                http://oneofthem.jp   
3                 http://www.0-6.com   
4  http://004gmbh.de/en/004-interact   

                                       category_list     status country_code  \
0                                              Media  operating          IND   
1  Application Platforms|Real Time|Social Network...  operating          USA   
2                                  Apps|Games|Mobile  operating          NaN   
3                                        Curated Web  operating          CHN   
4                                           Software  operating          USA   

  state_code                 region           city  founded_at  
0         16                 Mumbai         Mumbai         NaN  
1         DE             DE - Other  Delaware City  04-09-2014  
2        NaN                    NaN            NaN         NaN  
3         22                Beijing        Beijing  01-01-2007  
4         IL  Springfield, Illinois      Champaign  01-01-2010

In [5]:
# reading the rounds2 csv file
rounds2 = pd.read_csv('rounds2.csv',encoding='unicode_escape')

In [6]:
rounds2.head()

company_permalink  \
0                  /organization/-fame   
1               /ORGANIZATION/-QOUNTER   
2               /organization/-qounter   
3  /ORGANIZATION/-THE-ONE-OF-THEM-INC-   
4                /organization/0-6-com   

                           funding_round_permalink funding_round_type  \
0  /funding-round/9a01d05418af9f794eebff7ace91f638            venture   
1  /funding-round/22dacff496eb7acb2b901dec1dfe5633            venture   
2  /funding-round/b44fbb94153f6cdef13083530bb48030               seed   
3  /funding-round/650b8f704416801069bb178a1418776b            venture   
4  /funding-round/5727accaeaa57461bd22a9bdd945382d            venture   

  funding_round_code   funded_at  raised_amount_usd  
0                  B  05-01-2015         10000000.0  
1                  A  14-10-2014                NaN  
2                NaN  01-03-2014           700000.0  
3                  B  30-01-2014          3406878.0  
4                  A  19-03-2008          2000000.0

One thing to notice is the second row of both table companies and rounds2. The company permalink are '/Organization/-Qounter'and '/ORGANIZATION/-QOUNTER'. While these are the same, but they will be treated differently as <b>python is case sensitive</b>. So lets <b>lower case the permalinks in both the tables</b>.

In [7]:
# stripping to remove leading and trailing spaces if any and then lowercasing
companies['permalink'] = companies['permalink'].str.strip().str.lower()

In [8]:
# # stripping to remove leading and trailing spaces if any and then lowercasing
rounds2['company_permalink'] = rounds2['company_permalink'].str.strip().str.lower()

In [9]:
# getting a high level picture of the rounds2 dataframe -- 'raised_amount_usd' has some null values
rounds2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 114949 entries, 0 to 114948
Data columns (total 6 columns):
company_permalink          114949 non-null object
funding_round_permalink    114949 non-null object
funding_round_type         114949 non-null object
funding_round_code         31140 non-null object
funded_at                  114949 non-null object
raised_amount_usd          94959 non-null float64
dtypes: float64(1), object(5)
memory usage: 5.3+ MB


In [10]:
# looking if there any null values in the 'company_permalink' column in the rounds2 table
rounds2['company_permalink'].isnull().any()

False

Answer to Table 1.1, SI.No.1 -

In [11]:
# calculating the count of distinct entries in the column 'company_permalink'
rounds2['company_permalink'].nunique()

66370

In [12]:
# looking if there any null values in the 'permalink' column in the companies table
companies['permalink'].isnull().any()

False

Answer to Table 1.1, SI.No.2 -

In [13]:
# calculating the count of distinct entries in the column 'permalink'
companies['permalink'].nunique()

66368

Answer to Table 1.1, SI.No.3 -

In [14]:
# getting a high level picture of the rounds2 dataframe
companies.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 66368 entries, 0 to 66367
Data columns (total 10 columns):
permalink        66368 non-null object
name             66367 non-null object
homepage_url     61310 non-null object
category_list    63220 non-null object
status           66368 non-null object
country_code     59410 non-null object
state_code       57821 non-null object
region           58338 non-null object
city             58340 non-null object
founded_at       51147 non-null object
dtypes: object(10)
memory usage: 5.1+ MB


As the <b>total number of rows</b> in the dataframe 'companies' <b>are 66368</b> in number and <b>so are the number of unique entries in the column named 'permalink'</b>, hence this column can be used as the <b>unique key for each company</b>.

Answer to Table 1.1, SI.No.4

In [15]:
# innner joining both the dataframes
master_frame = pd.merge(left=rounds2,right=companies,how='inner',left_on='company_permalink',right_on='permalink')

In [16]:
master_frame.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 114942 entries, 0 to 114941
Data columns (total 16 columns):
company_permalink          114942 non-null object
funding_round_permalink    114942 non-null object
funding_round_type         114942 non-null object
funding_round_code         31139 non-null object
funded_at                  114942 non-null object
raised_amount_usd          94958 non-null float64
permalink                  114942 non-null object
name                       114941 non-null object
homepage_url               108810 non-null object
category_list              111535 non-null object
status                     114942 non-null object
country_code               106271 non-null object
state_code                 104003 non-null object
region                     104782 non-null object
city                       104785 non-null object
founded_at                 94422 non-null object
dtypes: float64(1), object(15)
memory usage: 14.9+ MB


As after using <b>inner join</b>, the <b>total number of rows have decreased</b>, that means <b>all the companies from rounds2 were not found in companies table</b>. That means that Yes, there are companies in rounds2 table which are not present in companies table.

In [17]:
master_frame.head()

company_permalink  \
0                  /organization/-fame   
1               /organization/-qounter   
2               /organization/-qounter   
3  /organization/-the-one-of-them-inc-   
4                /organization/0-6-com   

                           funding_round_permalink funding_round_type  \
0  /funding-round/9a01d05418af9f794eebff7ace91f638            venture   
1  /funding-round/22dacff496eb7acb2b901dec1dfe5633            venture   
2  /funding-round/b44fbb94153f6cdef13083530bb48030               seed   
3  /funding-round/650b8f704416801069bb178a1418776b            venture   
4  /funding-round/5727accaeaa57461bd22a9bdd945382d            venture   

  funding_round_code   funded_at  raised_amount_usd  \
0                  B  05-01-2015         10000000.0   
1                  A  14-10-2014                NaN   
2                NaN  01-03-2014           700000.0   
3                  B  30-01-2014          3406878.0   
4                  A  19-03-2008          2000000.0   

                             permalink                    name  \
0                  /organization/-fame                   #fame   
1               /organization/-qounter                :Qounter   
2               /organization/-qounter                :Qounter   
3  /organization/-the-one-of-them-inc-  (THE) ONE of THEM,Inc.   
4                /organization/0-6-com                 0-6.com   

             homepage_url                                      category_list  \
0      http://livfame.com                                              Media   
1  http://www.qounter.com  Application Platforms|Real Time|Social Network...   
2  http://www.qounter.com  Application Platforms|Real Time|Social Network...   
3     http://oneofthem.jp                                  Apps|Games|Mobile   
4      http://www.0-6.com                                        Curated Web   

      status country_code state_code      region           city  founded_at  
0  operating          IND         16      Mumbai         Mumbai         NaN  
1  operating          USA         DE  DE - Other  Delaware City  04-09-2014  
2  operating          USA         DE  DE - Other  Delaware City  04-09-2014  
3  operating          NaN        NaN         NaN            NaN         NaN  
4  operating          CHN         22     Beijing        Beijing  01-01-2007

In [18]:
master_frame['funding_round_type'].unique()

array(['venture', 'seed', 'undisclosed', 'equity_crowdfunding',
       'convertible_note', 'private_equity', 'debt_financing', 'angel',
       'grant', 'secondary_market', 'post_ipo_equity', 'post_ipo_debt',
       'product_crowdfunding', 'non_equity_assistance'], dtype=object)

In [19]:
#number of rows where raised amount is null
master_frame['raised_amount_usd'].isna().sum()

19984

In [20]:
# % such rows
(19984/114942)*100

17.38615997633589

In [21]:
# removing such rows from the analysis
master_frame = master_frame.dropna(axis=0, subset=['raised_amount_usd'])

In [22]:
# 19984 rows have been removed
master_frame.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 94958 entries, 0 to 114941
Data columns (total 16 columns):
company_permalink          94958 non-null object
funding_round_permalink    94958 non-null object
funding_round_type         94958 non-null object
funding_round_code         28169 non-null object
funded_at                  94958 non-null object
raised_amount_usd          94958 non-null float64
permalink                  94958 non-null object
name                       94957 non-null object
homepage_url               90626 non-null object
category_list              93914 non-null object
status                     94958 non-null object
country_code               89108 non-null object
state_code                 87355 non-null object
region                     87911 non-null object
city                       87914 non-null object
founded_at                 78992 non-null object
dtypes: float64(1), object(15)
memory usage: 12.3+ MB


In [23]:
# grouping the dataframe by the funding round type and then calculating the mean amount raised for each round
master_frame.groupby('funding_round_type').mean().round()

raised_amount_usd
funding_round_type                      
angel                           958694.0
convertible_note               1453439.0
debt_financing                17043526.0
equity_crowdfunding             538368.0
grant                          4300576.0
non_equity_assistance           411203.0
post_ipo_debt                168704572.0
post_ipo_equity               82182494.0
private_equity                73308593.0
product_crowdfunding           1363131.0
secondary_market              79649630.0
seed                            719780.0
undisclosed                   19242370.0
venture                       11748949.0

In [24]:
master_frame_venture = master_frame[master_frame['funding_round_type']=='venture']

In [25]:
master_frame_venture.head()

company_permalink  \
0                   /organization/-fame   
3   /organization/-the-one-of-them-inc-   
4                 /organization/0-6-com   
8   /organization/0ndine-biomedical-inc   
10                 /organization/0xdata   

                            funding_round_permalink funding_round_type  \
0   /funding-round/9a01d05418af9f794eebff7ace91f638            venture   
3   /funding-round/650b8f704416801069bb178a1418776b            venture   
4   /funding-round/5727accaeaa57461bd22a9bdd945382d            venture   
8   /funding-round/954b9499724b946ad8c396a57a5f3b72            venture   
10  /funding-round/3bb2ee4a2d89251a10aaa735b1180e44            venture   

   funding_round_code   funded_at  raised_amount_usd  \
0                   B  05-01-2015         10000000.0   
3                   B  30-01-2014          3406878.0   
4                   A  19-03-2008          2000000.0   
8                 NaN  21-12-2009           719491.0   
10                  B  09-11-2015         20000000.0   

                              permalink                    name  \
0                   /organization/-fame                   #fame   
3   /organization/-the-one-of-them-inc-  (THE) ONE of THEM,Inc.   
4                 /organization/0-6-com                 0-6.com   
8   /organization/0ndine-biomedical-inc  Ondine Biomedical Inc.   
10                 /organization/0xdata                  H2O.ai   

            homepage_url      category_list     status country_code  \
0     http://livfame.com              Media  operating          IND   
3    http://oneofthem.jp  Apps|Games|Mobile  operating          NaN   
4     http://www.0-6.com        Curated Web  operating          CHN   
8   http://ondinebio.com      Biotechnology  operating          CAN   
10        http://h2o.ai/          Analytics  operating          USA   

   state_code       region           city  founded_at  
0          16       Mumbai         Mumbai         NaN  
3         NaN          NaN            NaN         NaN  
4          22      Beijing        Beijing  01-01-2007  
8          BC    Vancouver      Vancouver  01-01-1997  
10         CA  SF Bay Area  Mountain View  01-01-2011

In [26]:
amount_raised_country_wise = master_frame_venture.groupby('country_code').sum()

In [27]:
amount_raised_country_wise.reset_index(inplace=True)

In [28]:
amount_raised_country_wise.head()

country_code  raised_amount_usd
0          ARE       3.426175e+08
1          ARG       3.143388e+08
2          ARM       2.651750e+07
3          AUS       1.322935e+09
4          AUT       6.309550e+08

In [29]:
top9 = amount_raised_country_wise.sort_values(by='raised_amount_usd',ascending=False).head(n=9)

In [30]:
top9

country_code  raised_amount_usd
94          USA       4.225108e+11
15          CHN       3.983542e+10
29          GBR       2.024563e+10
39          IND       1.439186e+10
12          CAN       9.583332e+09
28          FRA       7.259537e+09
42          ISR       6.907515e+09
21          DEU       6.346960e+09
45          JPN       3.363677e+09